[DOC-797] [2024.2.5] Ephemeral disk support and best practices (#28341)
yugabyte/yugabyte-db: docs/content/v2024.2/deploy/checklist.md

- Monitoring and alerts: Enable robust monitoring on the cluster so you are immediately alerted to node failures. YBA provides extensive monitoring and altering capabilities using a built-in Prometheus instance.

- Frequent backups: Increase your backup frequency when using ephemeral storage. YBA allows for incremental backups to occur as frequently as every 15 minutes. Also, consider backing up to a diversity of backup storage, and performing regular test restores.

#### Operational best practices and workflows

Working with ephemeral storage requires careful attention to standard operational workflows. The following are best practices for common scenarios.

##### Rolling restart of servers

The [Rolling restart](../../yugabyte-platform/manage-deployments/edit-config-flags/#batched-rolling-restart) universe action in YBA does not reboot the machine or unmount disks, so ephemeral data remains intact. You can perform rolling restarts as usual.

##### Operating System Patching and Node Reboots

> OS Patching procedures for self-managed clusters are customer-driven via scripts that call YBA APIs. See [Patch and upgrade the Linux operating system](../../yugabyte-platform/manage-deployments/upgrade-nodes/) for the recommended workflow.

Reviewer: ```suggestion
In YugabyteDB Anywhere, you perform OS patching for on-premises universes via scripts that call YugabyteDB Anywhere APIs. See [Patch and upgrade the Linux operating system](../../yugabyte-platform/manage-deployments/upgrade-nodes/) for the recommended workflow.
```

- Option 1: Blacklist the node Before Reboot (Drain): Blacklist the node, wait for draining to complete, reboot, and then remove the blacklist.
- Option 2: (Recommended) Remove and add node: Remove the node from the cluster, patch or rebuild its OS, and then add the node back to the universe.

> When using either approach, work on one node at a time and wait for the cluster to return to full replication of tablets based on their RF before moving to the next node.

Reviewer: ```suggestion

When using either approach, work on one node at a time and wait for the cluster to return to full replication of tablets based on their RF before moving to the next node.
```

- Option 2: (Recommended) Remove and add node: Remove the node from the cluster, patch or rebuild its OS, and then add the node back to the universe.
When using either approach, work on one node at a time and wait for the cluster to return to full replication of tablets based on their RF before moving to the next node.

##### Unplanned node outage (Failure scenario)

> In the event of an unplanned outage, assume the node's data is lost and proceed with a node replacement workflow (that is, using the Replace node universe action in YBA). If the node does come back on its own and starts to rejoin, monitor it closely.

Reviewer: ```suggestion
In the event of an unplanned outage, assume the node's data is lost and proceed with replacing the node (in YugabyteDB Anywhere, us the **Replace Node** action). If the node does come back on its own and starts to rejoin, monitor it closely.
```

##### Pause universe or full-cluster shutdown

> This operation is not supported or safe with ephemeral disks. Stopping all nodes means every node's local disk is wiped, resulting in total data loss. Never shut down all nodes simultaneously. YBA does not perform this action for universes created using the on-premises infrastructure provider. But you must ensure your own automation scripts also avoid this.

Reviewer: ```suggestion
This operation is not supported or safe with ephemeral disks. Stopping all nodes means every node's local disk is wiped, resulting in total data loss. Never shut down all nodes simultaneously.

YugabyteDB Anywhere does not perform this action for universes created using an on-premises provider. But you must ensure your own automation scripts also avoid this.
```

##### Software upgrades (YugabyteDB version upgrades)

> The [Upgrade Database Version](../../yugabyte-platform/manage-deployments/upgrade-software/) universe action in YBA does not reboot the machine or unmount disks. Therefore, this operation is safe. As per standard guidance during this operation, regardless of whether persistent or ephemeral disks are used, exercise caution in case a node fails during this operation as this may lead to a double fault scenario.

Reviewer: ```suggestion
In YugabyteDB Anywhere, the [Upgrade Database Version](../../yugabyte-platform/manage-deployments/upgrade-software/) option does not reboot the machine or unmount disks, and is therefore safe. As per standard guidance during this operation, regardless of whether persistent or ephemeral disks are used, exercise caution in case a node fails, as this may lead to a double fault scenario.
```

##### Scaling and adding/removing nodes

> - [Vertical scaling](../../explore/linear-scalability/horizontal-vs-vertical-scaling/#vertical-scale-up) (Instance type change to add memory, CPU or storage): Vertical scaling (that is, resizing a VM) for YBA on-premises universes is a manual operation.

Reviewer: ```suggestion
- [Vertical scaling](../../explore/linear-scalability/horizontal-vs-vertical-scaling/#vertical-scale-up) (Instance type change to add memory, CPU or storage): For YugabyteDB Anywhere on-premises universes, vertical scaling (resizing a VM) is a manual operation.
```

> You cannot detach and re-attach ephemeral disks to a new VM. Instead, perform a full move: add new nodes, and then remove the old ones.

Reviewer: ```suggestion
    You cannot detach and re-attach ephemeral disks to a new VM. Instead, perform a full move: add new nodes, and then remove the old ones.
```

  1. Add a new instance type to the on-premises provider for the new VM type.
  1. Add new nodes of this type to match the current universe count.
  1. Initiate an Edit universe operation to move from the old instance type to the new one. This triggers a full data migration to the new nodes.

  Note that as vertical scaling operations may introduce VM reboots on cloud providers, consider [horizontal scaling](../../explore/linear-scalability/horizontal-vs-vertical-scaling/#horizontal-scale-out) (scale out by adding nodes) instead of vertically scaling nodes.

- Disk scaling: Ephemeral disk sizes are fixed due to cloud provider limitations. To increase cluster capacity, add nodes rather than expanding disks.

- Removing nodes: Always blacklist nodes to safely drain data before removal. Note that ephemeral data is permanently lost when nodes are removed.

##### Additional tips

> - Automate: To reduce human error and streamline responses during situations like node outages, write scripts or use the YBA API to automate key workflows, such as adding, removing, and blacklisting nodes.

Reviewer: ```suggestion
- Automate: To reduce human error and streamline responses during situations like node outages, write scripts or use the YugabyteDB Anywhere API to automate key workflows, such as adding, removing, and blacklisting nodes.
```

- Operational complexity: Certain standard operations become dangerous with ephemeral disks. For instance, shutting down all nodes in a cluster with persistent disks is safe, but doing the same with an ephemeral disk cluster will result in total data loss. Operational workflows must be carefully reviewed to avoid these situations.

- Higher reliance on backups: Because the probability of a majority failure (such as a double fault in an RF3 configuration) is higher, the need to restore from a backup increases. High frequency backups are recommended; YugabyteDB Anywhere allows for incremental backups as frequently as every 15 minutes.

#### Universe design and configuration best practices

To safely and effectively use ephemeral disks with YugabyteDB, follow these design and configuration guidelines:

> - Have enough replication and redundancy: Always use a minimum replication factor (RF) of 3 for any production universe (and for all tablespaces for which you might override the RF) when using ephemeral disks. Due to the increased risk of "quorum-loss" scenarios, RF 5 or greater is recommended if your environment and performance tolerances allow. An RF 5 deployment can survive the failure of two fault domains without data loss (at the cost of more storage and write overhead).

Reviewer: ```suggestion
- Have enough replication and redundancy: Always use a minimum replication factor (RF) of 3 for any production universe (and for all tablespaces for which you might override the RF) when using ephemeral disks. Due to the increased risk of double fault scenarios, RF 5 or greater is recommended if your environment and performance tolerances allow. An RF 5 deployment can survive the failure of two fault domains without downtime (at the cost of more storage and write overhead).
```

- Distribute across zones: Spread nodes across availability zones so that a single zone outage won't result in all replicas of a tablet being dropped, causing data loss. Likewise, if you are using tablespaces to override replication settings for particular tables, spread each tablespace across multiple zones.

> - Persistent boot disk for OS and binaries: Configure YugabyteDB nodes so that the YugabyteDB software, configuration files, and logs reside on a persistent disk, and only the data directory is on the ephemeral disk. This design allows for quicker provisioning of nodes back into the cluster. It's also recommended to place logs on an independent persistent disk that's not the boot disk. Keeping logs persistent retains debugging information in case of a failure, and placing them on an independent, non-boot disk avoids the possibility of filling up the boot disk.

Reviewer: ```suggestion
- Persistent boot disk for OS and binaries: Configure YugabyteDB nodes so that the YugabyteDB software, configuration files, and logs reside on a persistent disk, and only the data directory is on the ephemeral disk. This design allows for quicker provisioning of nodes back into the cluster. In addition, put logs on an independent persistent disk that's not the boot disk. Keeping logs persistent retains debugging information in case of a failure, and placing them on an independent, non-boot disk avoids the possibility of filling up the boot disk.
```

  - Spare node rotation: If you maintain a free pool of spare nodes, be sure to apply [OS patches](../../yugabyte-platform/manage-deployments/upgrade-nodes/) to those spare VMs regularly.

- Monitoring and alerts: Enable robust monitoring on the cluster so you are immediately alerted to node failures. YugabyteDB Anywhere provides extensive monitoring and alerting using a built-in Prometheus instance.

- Frequent backups: Increase your backup frequency when using ephemeral storage. YugabyteDB Anywhere allows for incremental backups as frequently as every 15 minutes. Also, consider backing up to a diversity of backup storage, and performing regular test restores.

#### Operational best practices and workflows

Working with ephemeral storage requires careful attention to standard operational workflows. The following are best practices for common scenarios.

##### Rolling restart of servers

> [Rolling restarts](../../yugabyte-platform/manage-deployments/edit-config-flags/#batched-rolling-restart) universe action in YugabyteDB Anywhere do not reboot the machine or unmount disks, so ephemeral data remains intact. You can perform rolling restarts as usual.

Reviewer: ```suggestion
[Rolling restarts](../../yugabyte-platform/manage-deployments/edit-config-flags/#batched-rolling-restart) in YugabyteDB Anywhere do not reboot the machine or unmount disks, so ephemeral data remains intact. You can perform rolling restarts as usual.
```

Reviewer: Not sure why the link to enabling batched RR though

##### Operating system patching and node reboots

In YugabyteDB Anywhere, you perform OS patching for on-premises universes via scripts that call YugabyteDB Anywhere APIs. See [Patch and upgrade the Linux operating system](../../yugabyte-platform/manage-deployments/upgrade-nodes/) for the recommended workflow.

> When following the above documented guidance, if you must perform an OS reboot on a node, be sure to drain or remove it from the cluster first. The recommended approach is to remove and then re-add the node.

Reviewer: ```suggestion
When patching or performing maintenance, if you must reboot the OS on a node, be sure to drain or remove the node from the cluster first. The recommended approach is to remove and then re-add the node.
```

##### Unplanned node outage (Failure scenario)

> In the event of an unplanned outage, assume the node's data is lost and proceed with replacing the node (in YugabyteDB Anywhere, us the **Replace Node** action). If the node does come back on its own and starts to rejoin, monitor it closely.

Reviewer: ```suggestion
In the event of an unplanned outage, assume the node's data is lost and proceed with replacing the node (in YugabyteDB Anywhere, use the **Replace Node** action). If the node does come back on its own and starts to rejoin, monitor it closely.
```

##### Pause universe or full-cluster shutdown

This operation is not supported or safe with ephemeral disks. Stopping all nodes means every node's local disk is wiped, resulting in total data loss. Never shut down all nodes simultaneously.

YugabyteDB Anywhere does not perform this action for universes created using an on-premises provider. But you must ensure your own automation scripts also avoid this.

##### Software upgrades (YugabyteDB version upgrades)

In YugabyteDB Anywhere, the [Upgrade Database Version](../../yugabyte-platform/manage-deployments/upgrade-software/) option does not reboot the machine or unmount disks, and is therefore safe. As per standard guidance during this operation, regardless of whether persistent or ephemeral disks are used, exercise caution in case a node fails, as this may lead to a double fault scenario.

##### Scaling and adding/removing nodes

- [Vertical scaling](../../explore/linear-scalability/horizontal-vs-vertical-scaling/#vertical-scale-up) (Instance type change to add memory, CPU or storage): For YugabyteDB Anywhere on-premises universes, vertical scaling (resizing a VM) is a manual operation.

>     You cannot detach and re-attach ephemeral disks to a new VM. Instead, perform a full move: add new nodes, and then remove the old ones.

Reviewer: ```suggestion

    You cannot detach and re-attach ephemeral disks to a new VM. Instead, perform a full move: add new nodes, and then remove the old ones.
```
